listenbrainz/webserver/static/js/src/recent/RecentListens.tsx

```diff
@@ -0,0 +1,306 @@
+      try {
+        const data = await APIService.getFeedbackForUserForRecordings(
+          currentUser.name,
+          recordings
+        );
+        return data.feedback;
+      } catch (error) {
+        if (newAlert) {
+          newAlert(
+            "danger",
+            "Playback error",
+            typeof error === "object" ? error.message : error
+          );
+        }
+      }
+    }
+    return [];
+  };
+
+  loadFeedback = async () => {
+    const feedback = await this.getFeedback();
+    if (!feedback) {
+      return;
+    }
+    const recordingFeedbackMap: RecordingFeedbackMap = {};
+    feedback.forEach((fb: FeedbackResponse) => {
+      recordingFeedbackMap[fb.recording_msid] = fb.score;
+    });
+    this.setState({ recordingFeedbackMap });
+  };
+
+  updateFeedback = (
+    recordingMsid: string,
+    score: ListenFeedBack | RecommendationFeedBack
+  ) => {
+    const { recordingFeedbackMap } = this.state;
+    const newFeedbackMap = {
+      ...recordingFeedbackMap,
+      [recordingMsid]: score as ListenFeedBack,
+    };
+    this.setState({ recordingFeedbackMap: newFeedbackMap });
+  };
+
+  updateRecordingToPin = (recordingToPin: Listen) => {
+    this.setState({ recordingToPin });
+  };
+
+  updateRecordingToReview = (recordingToReview: Listen) => {
+    this.setState({ recordingToReview });
+  };
+
+  getFeedbackForRecordingMsid = (
+    recordingMsid?: string | null
+  ): ListenFeedBack => {
+    const { recordingFeedbackMap } = this.state;
+    return recordingMsid ? get(recordingFeedbackMap, recordingMsid, 0) : 0;
+  };
+
+  render() {
+    const { listens, recordingToPin, recordingToReview } = this.state;
+    const { newAlert } = this.props;
+    const { APIService, currentUser } = this.context;
+
+    return (
+      <div role="main">
+        <h3>Recent listens</h3>
+        <div className="row">
+          <div className="col-md-8">
+            {!listens.length && (
+              <div className="lead text-center">
+                <p>No listens yet</p>
+              </div>
+            )}
+            {listens.length > 0 && (
+              <>
+                <div id="listens">
+                  {listens.map((listen) => {
+                    const recordingMBID = getRecordingMBID(listen);
+                    const artistMBIDs = getArtistMBIDs(listen);
+                    const trackMBID = get(
+                      listen,
+                      "track_metadata.additional_info.track_mbid"
+                    );
+                    const releaseGroupMBID = getReleaseGroupMBID(listen);
+
+                    const isListenReviewable =
+                      Boolean(recordingMBID) ||
+                      artistMBIDs?.length ||
+                      Boolean(trackMBID) ||
+                      Boolean(releaseGroupMBID);
+                    /* eslint-disable react/jsx-no-bind */
+                    const additionalMenuItems = (
+                      <>
+                        <ListenControl
```

Comment:
Listens cannot be pinned unless they have either msid or mbid. So, a now playing listen without mbids submitted by the user cannot be pinned. So we should check that here.

Reply:
I realized that since this is RecentListens page, now playing listens won't appear here. But I'll probably get confused in future again about this. A comment mentioning why we don't need the check would be nice.

Reply:
Also, this comment is probably still relevant to Listens.tsx see Line 716 in that file.